src/ArrowFunctions.cpp

```diff
@@ -50,68 +50,127 @@ bool check_status_ok(arrow::Status status, char** errMsg) {
+    else if(myType->id() == arrow::Type::FLOAT)
+      return ARROWFLOAT;
+    else if(myType->id() == arrow::Type::DOUBLE)
+      return ARROWDOUBLE;
+    else // TODO: error type not supported
+      return ARROWUNDEFINED;
+  } catch (const std::exception& e) {
+    *errMsg = strdup(e.what());
     return ARROWERROR;
   }
-  auto myType = sc -> field(idx) -> type();
-
-  if(myType->id() == arrow::Type::INT64)
-    return ARROWINT64;
-  else if(myType->id() == arrow::Type::INT32)
-    return ARROWINT32;
-  else if(myType->id() == arrow::Type::UINT64)
-    return ARROWUINT64;
-  else if(myType->id() == arrow::Type::UINT32)
-    return ARROWUINT32;
-  else if(myType->id() == arrow::Type::TIMESTAMP)
-    return ARROWTIMESTAMP;
-  else if(myType->id() == arrow::Type::BOOL)
-    return ARROWBOOLEAN;
-  else if(myType->id() == arrow::Type::STRING)
-    return ARROWSTRING;
-  else if(myType->id() == arrow::Type::FLOAT)
-    return ARROWFLOAT;
-  else if(myType->id() == arrow::Type::DOUBLE)
-    return ARROWDOUBLE;
-  else // TODO: error type not supported
-    return ARROWUNDEFINED;
 }
 
 int cpp_getStringColumnNumBytes(const char* filename, const char* colname, void* chpl_offsets, char** errMsg) {
-  int64_t ty = cpp_getType(filename, colname, errMsg);
-  auto offsets = (int64_t*)chpl_offsets;
+  try {
+    int64_t ty = cpp_getType(filename, colname, errMsg);
+    auto offsets = (int64_t*)chpl_offsets;
+
+    if(ty == ARROWSTRING) {
+      int size = 0;
+      std::unique_ptr<parquet::ParquetFileReader> parquet_reader =
+        parquet::ParquetFileReader::OpenFile(filename, false);
+
+      std::shared_ptr<parquet::FileMetaData> file_metadata = parquet_reader->metadata();
+      int num_row_groups = file_metadata->num_row_groups();
+
+      int64_t i = 0;
+      for (int r = 0; r < num_row_groups; r++) {
+        std::shared_ptr<parquet::RowGroupReader> row_group_reader =
+          parquet_reader->RowGroup(r);
 
-  if(ty == ARROWSTRING) {
-    int size = 0;
+        int64_t values_read = 0;
+
+        std::shared_ptr<parquet::ColumnReader> column_reader;
+
+        auto idx = file_metadata -> schema() -> ColumnIndex(colname);
+
+        if(idx < 0) {
+          std::string dname(colname);
+          std::string fname(filename);
+          std::string msg = "Dataset: " + dname + " does not exist in file: " + fname; 
+          *errMsg = strdup(msg.c_str());
+          return ARROWERROR;
+        }
+        column_reader = row_group_reader->Column(idx);
+        parquet::ByteArrayReader* ba_reader =
+          static_cast<parquet::ByteArrayReader*>(column_reader.get());
+
+        while (ba_reader->HasNext()) {
+          parquet::ByteArray value;
+          (void)ba_reader->ReadBatch(1, nullptr, nullptr, &value, &values_read);
+          size += value.len + 1; // add 1 to account for zero that will be added in Chapel array
+          offsets[i++] = value.len + 1;
+        }
+      }
+      return size;
+    }
+    return ARROWUNDEFINED;
```

Comment:
Would we want an error in the event the something is not a string here? Returning undefined does not seem wrong, but if the type does not meet the criteria I would think we would want an error. I may be wrong in my thought process here.

Reply:
In the case where this function is called and it isn't a string, I think it should be an error, but in order for that to happen something would have to go pretty wrong on the server side, so it is pretty unlikely. The `ARROWUNDEFINED` and `ARROWERROR` are actually the same thing, so maybe we should just condense them to be only `ARROWERROR` to remove some confusion. I don't think having the second variable is all that valuable.

Reply:
That make sense. I was not sure if that was the intended functionality or not, which is why I brought it up. I definitely agree that it would make sense to only use `ARROWERROR` to avoid confusion.